### ISSUE TITLE
Converting lists of array-like objects to numpy arrays can be very slow

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#8562]
NumPy seems to attempt to iterate instead of calling `__array__`.

Consider:
```python
class ArrayLike(object):
  def __init__(self, array):
    self.array = array

  def __len__(self):
    return len(self.array)
  
  def __iter__(self):
    print('calling __iter__')
    return iter(self.array)
  
  def __getitem__(self, index):
    print('calling __getitem__ with index={}'.format(index))
    return self.array[index]

  def __array__(self, dtype=None):
    print('calling __array__')
    return np.asarray(self.array, dtype=dtype)
```

```
>>> a = ArrayLike(np.arange(3))

>>> np.array(a)
calling __array__
array([0, 1, 2])

>>> list(a)
calling __iter__
[0, 1, 2]

>>> np.array([a])
calling __array__
calling __iter__
calling __iter__
array([[0, 1, 2]])
```

So actually, it looks like NumPy calls `__array__`, and then `__iter__` twice!

Is there a good reason for this behavior? If not, maybe we can fix it? Either way, a pointer to the relevant place in the codebase would be appreciated.

This causes trouble for xarray users (https://github.com/pydata/xarray/issues/1247) because iterating over xarray objects creates new xarray objects, which takes ~100 us each. That adds up fast for large arrays.

**Comment by seberg** [comment]
Its all in ctors.c, called by PyArray_FromAny, and discover_dimensions is probably a big part of it (especially the dimension wise iteration may be due to it).

**Comment by wmbr** [comment]
I also stumbled upon this strange behavior in the following example:
```python
class DummyArray:
	def __array__(self):
		print('calling __array__')
		return np.zeros((1,))
```
```

>>> a = DummyArray()
>>> np.array([a])
calling __array__
TypeError: float() argument must be a string or a number, not 'DummyArray'
```

So somehow `__array__` is called but then the result is not used?

I tried digging into ctors.c and ended up at PyArray_DTypeFromObject in common.c which looks like it should be doing the right thing, so I don't see where this behaviour stems from. But I admit I only understand half of that C code, so I hope someone with more insight can help us here.
